README
脚本的运行的前提，需要打开libvirt的监听

centos修改/etc/init.d/libvirtd

  LIBVIRTD_ARGS=-l

ubuntu下如下设置：
    openstack的官网文档中有介绍：
    /etc/libvirt/libvirt.conf:

    before  : #listen_tls = 0
    after : listen_tls = 0

    before : #listen_tcp = 1
    afterfter : listen_tcp = 1

    add: auth_tcp = "none"


    listen_addr = "localhost"  #这个是我们加的，仅仅监听本机

    Modify /etc/init/libvirt-bin.conf

    before : exec /usr/sbin/libvirtd  -d
    after : exec /usr/sbin/libvirtd -d -l

      Modify /etc/default/libvirt-bin

    before :libvirtd_opts=" -d"
    after :libvirtibvirtd_opts=" -d -l"

then, restart libvirt. Make sure libvirt is restarted.

# stop libvirt-bin && start libvirt-bin
# ps -ef | grep libvirt
root 1145 1 0 Nov27 ? 00:00:03 /usr/sbin/libvirtd -d -l




数据库的表结构：


desc instance;
+-------------+-----------+------+-----+---------+----------------+
| Field       | Type      | Null | Key | Default | Extra          |
+-------------+-----------+------+-----+---------+----------------+
| id          | int(11)   | NO   | PRI | NULL    | auto_increment |
| instance_id | int(7)    | YES  |     | NULL    |                |
| timestamp   | char(15)  | YES  |     | NULL    |                |
| cpu         | char(5)   | YES  |     | NULL    |                |
| mem         | char(5)   | YES  |     | NULL    |                |
| disk        | char(100) | YES  |     | NULL    |                |
| net         | char(100) | YES  |     | NULL    |                |
+-------------+-----------+------+-----+---------+----------------+
7 rows in set (0.00 sec)


脚本输出的对应关系：

('264','1329187275.52','0.249','10.34',' vda-0.206-0.0-0.0-37580963840::vdb-0.004-0.0-0.0-85899345920::',' vnet12-468-0-3288953-0::vnet13-0-0-36090-0::')
('instance_id','timestamp','cpu_usage','memory_usage','disk_name-disk_usage-read_speed-write_speed-disk_size::','netcard_name-total_upload-upload_speed-total_download-download_speed')

unit of disk:Kb/s
ubit of net :Kb/s

'''


changelog
取消守护进程后台运行
	需要在crontab中设置


monitor_vm.py
#!/usr/bin/env python2.7
#coding= utf8

import libvirt
import time
import Queue,threading
import os,sys
import MySQLdb
import traceback
import conf
from  my_libvirt_lib import *

queue=Queue.Queue()
MAXTHREAD=5
log_dir=sys.path[0]

uri='qemu:///system'
def get_node_info():
	global MAXTHREAD
	conn=libvirt.open(uri)
	doms=conn.listDomainsID()
	for dom_id in doms:
		queue.put(dom_id)
		class get_dom_info(threading.Thread):
			global queue
			def run(self):
				while not queue.empty():
					dom_id=queue.get()
					vm_status='down'
					block_status0={}
					block_status1={}
					net_status0={}
					net_status1={}
					block_info=[]
					block_status=' '
					nic_status=' '
					try:
						dom=conn.lookupByID(dom_id)
						dom_info=dom.info()
						disks=get_devices(dom,"devices/disk/target","dev")
						nics=get_devices(dom,"devices/interface/target","dev")
						for block in disks:
							block_status0[block]=dom.blockStats(block)
						for nic in nics:
							net_status0[nic]=[os.popen("cat /proc/net/dev |grep -w '"+nic+"' |awk '{print $10}'").readlines()[0][:-1],os.popen("cat /proc/net/dev |grep '"+nic+"' |awk '{print $2}'").readlines()[0][:-1]]
						Lstart_time=time.time()*1000000000
						Dstart_time=dom.info()[4]
						time.sleep(2)
						Dstop_time=dom.info()[4]
						Lstop_time=time.time()*1000000000
						for block in disks:
							block_status1[block]=dom.blockStats(block)
						for nic in nics:
							net_status1[nic]=[os.popen("cat /proc/net/dev |grep -w  '"+nic+"' |awk '{print $10}'").readlines()[0][:-1],os.popen("cat /proc/net/dev |grep '"+nic+"' |awk '{print $2}'").readlines()[0][:-1]]
						for block in get_devices(dom,"devices/disk/source","file"):
							block_info.append(dom.blockInfo(block,0))
						for nic in nics:
							try:
								nic_status+=nic+"-"+str(int(net_status1[nic][1]))+"-"+str((int(net_status1[nic][1])-int(net_status0[nic][1]))/2048)+"-"+str(int(net_status1[nic][0]))+"-"+str((int(net_status1[nic][0])-int(net_status0[nic][0]))/2048)+"::"
							except:
								nic_stauus=''
						i=0
						for block in get_devices(dom,"devices/disk/target","dev"):
							block_status+=block+"-"+str(block_info[i][1]/1.0/block_info[i][0]*100)[:5]+"-"+str((block_status1[block][1]-block_status0[block][1])/2048.0)+"-"+str((block_status1[block][3]-block_status0[block][3])/2048.0)+"-"+str(block_info[i][0])+"::"
							i=i+1
						cpu_usage=(Dstop_time-Dstart_time)/(Lstop_time-Lstart_time)/int(dom.info()[3])*100
						if cpu_usage>100:
							cpu_usage=100
						if dom_info[0]==1:
							vm_status='active'
						memstatus=os.popen("awk /VmRSS/'{print $2}' /proc/`ps aux | grep -v 'grep' |grep "+dom.name()+"|awk '{print $2}' `/status").readlines()[0][:-1]
						memusage='%.2f' %  (int(memstatus)*100.0/int(dom_info[2]))
						if memusage>100:
							memusage=100
						name=ElementTree.fromstring(dom.XMLDesc(0)).findtext('uuid')
					except (KeyboardInterrupt,SystemExit):
						raise
					except:
						traceback.print_exc()
						break
					db_conn=MySQLdb.connect(host=conf.db_host,user=conf.db_user,passwd=conf.db_passwd,db=conf.db)
					cursor=db_conn.cursor()
					cmd="insert into instance (instance_id,timestamp,cpu,mem,disk,net)  values ('"+str(name)+"','"+str(time.time())+"','"+str(cpu_usage)[:5]+"','"+str(memusage)+"','"+block_status+"','"+nic_status+"')"
					print cmd
					cursor.execute(cmd)
					db_conn.commit()
					db_conn.close()
					queue.task_done()

	if queue.qsize()<MAXTHREAD:
		MAXTHREAD=queue.qsize()
	for i in xrange(0,MAXTHREAD):
		get_dom_info().start()
	queue.join()

if __name__=="__main__":
	get_node_info()


my_libvirt_lib.py
# vim: tabstop=4 shiftwidth=4 softtabstop=4
import libvirt,os,sys
from xml.etree import ElementTree

def get_devices(dom,path,devs):
	tree=ElementTree.fromstring(dom.XMLDesc(0))
	devices=[]
	for target in tree.findall(path):
		dev=target.get(devs)
		if not dev in devices:
			devices.append(dev)
	return devices
def daemonize(stdout='/dev/null', stderr=None, stdin='/dev/null',pidfile=None ):
       
	try:
		pid = os.fork()
		if pid > 0: sys.exit(0) 
	except OSError, e:
		sys.stderr.write("fork #1 failed: (%d) %s\n" % (e.errno, e.strerror))
		sys.exit(1)
	
	os.chdir("/")
	os.umask(0)
	os.setsid()

	try:
		pid = os.fork()
		if pid > 0: sys.exit(0) 
	except OSError, e:
		print 'second fork error'
		sys.stderr.write("fork #2 failed: (%d) %s\n" % (e.errno, e.strerror))
		sys.exit(1)
	if not stderr: stderr = stdout
	si = file(stdin, 'r')
	so = file(stdout, 'w')
	se = file(stderr, 'a+', 0)
	pid = str(os.getpid())
	print "Start with Pid: %s\n"  % pid
	sys.stderr.flush()
	if pidfile: file(pidfile,'w').write("%s\n" % pid)
	sys.stdout.flush()
	sys.stderr.flush()
	os.dup2(si.fileno(), sys.stdin.fileno())
	os.dup2(so.fileno(), sys.stdout.fileno())
	os.dup2(se.fileno(), sys.stderr.fileno())


instance.py
#!/usr/bin/env python2.7
#coding= utf8

import libvirt
import time
import Queue,threading
import os,sys
import MySQLdb
import traceback
import conf
from  my_libvirt_lib import *

queue=Queue.Queue()
MAXTHREAD=5
log_dir=sys.path[0]

uri='qemu+tcp:///system'
def get_node_info():
	global MAXTHREAD
	conn=libvirt.open(uri)
	doms=conn.listDomainsID()
	if doms==[]:
		while True:
			doms=conn.listDomainsID()
			if doms==[]:
				time.sleep(60)
	for dom_id in doms:
		queue.put(dom_id)
		class get_dom_info(threading.Thread):
			global queue
			def run(self):
				while not queue.empty():
					dom_id=queue.get()
					vm_status='down'
					block_status0={}
					block_status1={}
					net_status0={}
					net_status1={}
					block_info=[]
					block_status=' '
					nic_status=' '
					try:
						dom=conn.lookupByID(dom_id)
						dom_info=dom.info()
						disks=get_devices(dom,"devices/disk/target","dev")
						nics=get_devices(dom,"devices/interface/target","dev")
						for block in disks:
							block_status0[block]=dom.blockStats(block)
						for nic in nics:
							net_status0[nic]=[os.popen("cat /proc/net/dev |grep '"+nic+"' |awk '{print $10}'").readlines()[0][:-1],os.popen("cat /proc/net/dev |grep '"+nic+"' |awk '{print $2}'").readlines()[0][:-1]]
						Lstart_time=time.time()*1000000000
						Dstart_time=dom.info()[4]
						time.sleep(2)
						Dstop_time=dom.info()[4]
						Lstop_time=time.time()*1000000000
						for block in disks:
							block_status1[block]=dom.blockStats(block)
						for nic in nics:
							net_status1[nic]=[os.popen("cat /proc/net/dev |grep '"+nic+"' |awk '{print $10}'").readlines()[0][:-1],os.popen("cat /proc/net/dev |grep '"+nic+"' |awk '{print $2}'").readlines()[0][:-1]]
						for block in get_devices(dom,"devices/disk/source","file"):
							block_info.append(dom.blockInfo(block,0))
						for nic in nics:
							try:
								nic_status+=nic+"-"+str(int(net_status1[nic][1]))+"-"+str((int(net_status1[nic][1])-int(net_status0[nic][1]))*4)+"-"+str(int(net_status1[nic][0]))+"-"+str((int(net_status1[nic][0])-int(net_status0[nic][0]))*4)+"::"
							except:
								nic_stauus=''
						i=0
						for block in get_devices(dom,"devices/disk/target","dev"):
							block_status+=block+"-"+str(block_info[i][1]/1.0/block_info[i][0]*100)[:5]+"-"+str((block_status1[block][1]-block_status0[block][1])/2048.0)+"-"+str((block_status1[block][3]-block_status0[block][3])/2048.0)+"-"+str(block_info[i][0])+"::"
							i=i+1
						cpu_usage=(Dstop_time-Dstart_time)/(Lstop_time-Lstart_time)/int(dom.info()[3])*100
						if cpu_usage>100:
							cpu_usage=100
						if dom_info[0]==1:
							vm_status='active'
						memstatus=os.popen("awk /VmRSS/'{print $2}' /proc/`ps aux | grep -v 'grep' |grep "+dom.name()+"|awk '{print $2}' `/status").readlines()[0][:-1]
						memusage=int(memstatus)*100.0/int(dom_info[2])
						if memusage>100:
							memusage=100
						name=int(dom.name().split('-')[-1],16)
					except (KeyboardInterrupt,SystemExit):
						raise
					except:
						traceback.print_exc()
						break
					db_conn=MySQLdb.connect(host=conf.db_host,user=conf.db_user,passwd=conf.db_passwd,db=conf.db)
					cursor=db_conn.cursor()
					cmd="insert into instance (instance_id,timestamp,cpu,mem,disk,net)  values ('"+str(name)+"','"+str(time.time())+"','"+str(cpu_usage)[:5]+"','"+str(memusage)+"','"+block_status+"','"+nic_status+"')"
					print cmd
					cursor.execute(cmd)
					db_conn.close()	
					queue.task_done()
	
	if queue.qsize()<MAXTHREAD:
		MAXTHREAD=queue.qsize()
	for i in xrange(0,MAXTHREAD):
		get_dom_info().start()
	queue.join()
def start():
	while True:
		get_node_info()
		time.sleep(300)

if __name__=="__main__":
	daemonize(stdout=log_dir+'/instancestdout.log', stderr=log_dir+'/instancestderr.log',pidfile=log_dir+'/monitor/pid.txt')
	start()
